### PR TITLE
Pointer overflow checks should detect overflow in offset multiplication

### LINKED ISSUE #6631: Corner case in pointer-overflow-check
CBMC version: 5.48
Operating system: OSX
Exact command line resulting in the issue: `cbmc --pointer-overflow-check test.c `
What behaviour did you expect: Overflow detected
What happened instead:
```
╰─$ cbmc --pointer-overflow-check test.c
CBMC version 5.48.0 (cbmc-5.48.0) 64-bit x86_64 macos
Parsing test.c
Converting
Type-checking test
Generating GOTO Program
Adding CPROVER library (x86_64)
Removal of function pointers and virtual functions
Generic Property Instrumentation
Running with 8 object bits, 56 offset bits (default)
Starting Bounded Model Checking
Runtime Symex: 0.00214797s
size of program expression: 45 steps
simple slicing removed 0 assignments
Generated 2 VCC(s), 0 remaining after simplification
Runtime Postprocess Equation: 2.109e-05s

** Results:
test.c function main
[main.pointer_arithmetic.1] line 7 pointer arithmetic: dead object in p + 9223372036854775807l: SUCCESS
[main.pointer_arithmetic.2] line 7 pointer arithmetic: pointer outside object bounds in p + 9223372036854775807l: SUCCESS

** 0 of 2 failed (1 iterations)
VERIFICATION SUCCESSFUL
```

```c
#include <limits.h>
#include <stdint.h>

int main() {
    int i[5];
    int*p = &i[1]; // Spurious success. Change to any other offset to get the expected fail
    int*q = p + SSIZE_MAX;
}
```

### PR DESCRIPTION
**Pull request by tautschnig**
Pointer arithmetic requires multiplication of the offset by the size of
the base type (for any base type larger than 1 byte). Such a
multiplication isn't introduced until the back-end, where no opportunity
for adding properties exists anymore. Therefore, synthesize the
multiplication to generate arithmetic overflow checks at the GOTO level.

Fixes: #6631

- [x] Each commit message has a non-empty body, explaining why the change was made.
- n/a Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- n/a The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [x] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- n/a My commit message includes data points confirming performance improvements (if claimed).
- [ ] My PR is restricted to a single feature or bugfix.
- n/a White-space or formatting changes outside the feature-related changed lines are in commits of their own.

### CODE REVIEW

File: src/analyses/goto_check_c.cpp [hunk 1]
```diff
@@ -340,6 +340,82 @@ class goto_check_ct
...
 };
 
+/// Allows to:
+/// - override a Boolean flag with a new value via `set_flag`
+/// - set a Boolean flag to false via `disable_flag`, such that
+///   previous `set_flag` are overridden and future `set_flag` are ignored.
+///
+/// A flag's initial value (before any `set_flag` or `disable_flag`) is restored
+/// when the entire object goes out of scope.
+class flag_resett
```
Comment (chris-ryder) [inline src/analyses/goto_check_c.cpp]: I realise this is just code you've moved, but just wondering about the naming of this class (sorry...) Conceptually what this class is really about is less about "resetting" a flag, and more about introducing a "scope" for the flag? So maybe call it something like `flag_scopet` ? or `flag_overridet` ? One for a later PR though.

**Review by chris-ryder — APPROVED** [review]

File: src/analyses/goto_check_c.cpp (hunk 1 above)
Comment (tautschnig) [reply src/analyses/goto_check_c.cpp]: Good call, I think `flag_overridet` is the best match for what it does. I've inserted a commit to do the renaming.

**Review by chris-ryder — APPROVED** [review]
Re-asserting my approval after the extra rename commit, thanks :-)
